Extract libraries

In [4]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

Extract and preparation data

In [5]:
data_negative = pd.read_csv('data/processedNegative.csv')
data_neutral = pd.read_csv('data/processedNeutral.csv')
data_positive = pd.read_csv('data/processedPositive.csv')

In [6]:
data_negative.head(5)

,How unhappy some dogs like it though,talking to my over driver about where I'm goinghe said he'd love to go to New York too but since Trump it's probably not,Does anybody know if the Rand's likely to fall against the dollar? I got some money I need to change into R but it keeps getting stronger unhappy,I miss going to gigs in Liverpool unhappy,There isnt a new Riverdale tonight ? unhappy,it's that A*dy guy from pop Asia and then the translator so they'll probs go with them around Aus unhappy,Who's that chair you're sitting in? Is this how I find out. Everyone knows now. You've shamed me in pu,don't like how jittery caffeine makes me sad,My area's not on the list unhappy think I'll go LibDems anyway,I want fun plans this weekend unhappy,...,and yet if parents invest in child's emotional education by taking child out of school on holiday early that's un,YG should have sent them to MCD. I want to see them holding the trophy unhappy anyways .9,i want more orientation unhappy,unhappy they not,YG should have sent them to MCD. I want to see them holding the trophy unhappy anyways .10,wish knock out lang talaga for the new school year are good and cooperative groupmates please unhappy,i miss so much unhappy,Same unhappy .1,Hi instant message your friend friend lang,hindi close friend? unhappy


In [7]:
data_neutral.head(5)

,Pak PM survives removal scare,but court orders further probe into corruption charge.,Supreme Court quashes criminal complaint against cricketer for allegedly depicting himself as on magazine cover.,Art of Living's fights back over Yamuna floodplain damage,livid.,FCRA slap on NGO for lobbying...But was it doing so as part of govt campaign?,Why doctors,pharma companies are opposing names on,Why a bicycle and not a CM asked. His officer learnt ground reality -- and a dip in a river.,It's 2017. making law to ban And MHA is sitting on draft.,...,Payments Bank to offer 4.5% on deposits up to Rs25000,5% on Rs 25000-50000,5.5% on Rs 50000-100000.,Supreme Court shoots down govt bid to put sports ministry bureaucrat in panel to run,Historian Ram Guha,IDFC official Vikram Limaye,former captain Diana Edulji are others in panel to run,Supreme Court names former CAG as head of 4-member panel of administrators to run 1/3,Court summons CM suspended BJP MP as accused in criminal defamation plea by,Amulya Patnaik has been appointed new Delhi police commissioner. Patnaik is a 1985 AGMUT cadre IPS officer.


In [8]:
data_positive.head(5)

,An inspiration in all aspects: Fashion,fitness,beauty and personality. :)KISSES TheFashionIcon,Apka Apna Awam Ka Channel Frankline Tv Aam Admi Production Please Visit Or Likes Share :)Fb Page :...,Beautiful album from the greatest unsung guitar genius of our time - and I've met the great backstage,Good luck to Rich riding for great project in this Sunday. Can you donate?,Omg he... kissed... him crying with joy,happy anniv ming and papi!!!!! love love happy,thanks happy,C'mon Tweeps,...,$ES_F $SPY Bulls are just relentless happy Some setups I like today,Thanks for the recent follow Happy to connect happy have a great Thursday..33,Thanks for the recent follow Happy to connect happy have a great Thursday. Want this.5,Thanks for the recent follow Happy to connect happy have a great Thursday. Want this.7,you know that the problem still exist :D,Thanks for the recent follow Happy to connect happy have a great Thursday. Get this.1,- top engaged members this week happy,ngam to weeks left for cadet pilot exam crying with joy,Great! You're welcome Josh happy ^Adam,Sixth spot not applicable Team! Higher pa! :)KISSES TheFashionIcon.4


In [6]:
data_negative = data_negative.T
data_neutral = data_neutral.T
data_positive = data_positive.T

In [95]:
print(data_negative.shape)
print(data_neutral.shape)
print(data_positive.shape)

(1117, 0)
(1570, 0)
(1186, 0)


In [7]:
data_negative['sentement'] = 'negative'
data_neutral['sentement'] = 'neutral'
data_positive['sentement'] = 'positive'

In [8]:
data_negative.index.name = 'sentence'
data_neutral.index.name = 'sentence'
data_positive.index.name = 'sentence'

In [98]:
data_positive.head(5)

,sentement
sentence,
An inspiration in all aspects: Fashion,positive
fitness,positive
beauty and personality. :)KISSES TheFashionIcon,positive
Apka Apna Awam Ka Channel Frankline Tv Aam Admi Production Please Visit Or Likes Share :)Fb Page :...,positive
Beautiful album from the greatest unsung guitar genius of our time - and I've met the great backstage,positive


In [9]:
df = pd.concat([data_negative, data_neutral, data_positive])

In [10]:
import re

leak_words = ['happy', 'unhappy']

def remove_leak(text):
    text = str(text)
    for w in leak_words:
        text = re.sub(rf'\b{w}\b', '', text, flags=re.IGNORECASE)
    return text.strip()

df.index = df.index.map(remove_leak)

In [101]:
df.sample(5)

,sentement
sentence,
if i get 900 retweets and 900 likes. help me :(un.2,negative
we don%27t talk anymore like we used to do,negative
Why may not show in,neutral
Superstar thanks so much for the kind words! Look forward to seeing you soon buddy smile.3,positive
I dohn't want toh cohmment ohn it: Amit Shah oh,neutral


In [102]:
df.shape

(3873, 1)

Splitting (80/20, stratified by 'sentement')

In [18]:
train, test = train_test_split(df, test_size=0.2, stratify=df['sentement'])

In [104]:
print(test.head(3))
print(test.head(3))
print(' ')
print(f'Train size: {train.shape}')
print(f'Test size: {test.shape}')

                                                   sentement
sentence                                                    
Your titties are fantastic. Kissy Kissy             positive
I love it when Ben sleeps in because I get to w...  positive
How the formidable Madame of Calcutta's runs he...   neutral
                                                   sentement
sentence                                                    
Your titties are fantastic. Kissy Kissy             positive
I love it when Ben sleeps in because I get to w...  positive
How the formidable Madame of Calcutta's runs he...   neutral
 
Train size: (3098, 1)
Test size: (775, 1)


Our plan of 21 different experiments:

|          | 0/1 (if word exists) | word counts | TFIDF |
|---|---|---|---|
| just tokenization | | | |
| stemming | | | |
| lemmatization | | | |
| stemming + misspellings | | | |
| lemmatization + misspellings | | | |
| stemming + misspellings + stop-words | | | |
| lemmatization + misspellings + stop-words | | | |

In [19]:
X_train = train.index.astype(str)
y_train = train['sentement']
X_test  = test.index.astype(str)
y_test  = test['sentement']

1. **Just tokenization**

In [106]:
vectorizer = CountVectorizer(binary=True)
X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec  = vectorizer.transform(X_test)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_vec, y_train)
preds = model.predict(X_test_vec)

print("0/1:", accuracy_score(y_test, preds))

0/1: 0.7651612903225806


In [107]:
vectorizer = CountVectorizer()
X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec  = vectorizer.transform(X_test)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_vec, y_train)
preds = model.predict(X_test_vec)

print("word counts:", accuracy_score(y_test, preds))

word counts: 0.7574193548387097


In [108]:
vectorizer = TfidfVectorizer()
X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec  = vectorizer.transform(X_test)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_vec, y_train)
preds = model.predict(X_test_vec)

print("TFIDF:", accuracy_score(y_test, preds))

TFIDF: 0.7612903225806451


2. **Stemming**

In [24]:
import nltk
from nltk.stem import PorterStemmer
from nltk.tokenize import word_tokenize

nltk.download('punkt')
nltk.download('punkt_tab')

stemmer = PorterStemmer()

[nltk_data] Downloading package punkt to /Users/ggang/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to /Users/ggang/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


In [21]:
def apply_stemming(text):
    words = word_tokenize(text)
    stemmed = [stemmer.stem(w) for w in words]
    return ' '.join(stemmed)

In [111]:
X_train_stem = X_train.map(apply_stemming)
X_test_stem  = X_test.map(apply_stemming)

In [112]:
vectorizer = CountVectorizer(binary=True)
X_train_stem_vec = vectorizer.fit_transform(X_train_stem)
X_test_stem_vec  = vectorizer.transform(X_test_stem)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_stem_vec, y_train)
preds = model.predict(X_test_stem_vec)

print("0/1:", accuracy_score(y_test, preds))

0/1: 0.7625806451612903


In [113]:
vectorizer = CountVectorizer()
X_train_stem_vec = vectorizer.fit_transform(X_train_stem)
X_test_stem_vec  = vectorizer.transform(X_test_stem)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_stem_vec, y_train)
preds = model.predict(X_test_stem_vec)

print("word counts:", accuracy_score(y_test, preds))

word counts: 0.7664516129032258


In [114]:
vectorizer = TfidfVectorizer()
X_train_stem_vec = vectorizer.fit_transform(X_train_stem)
X_test_stem_vec  = vectorizer.transform(X_test_stem)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_stem_vec, y_train)
preds = model.predict(X_test_stem_vec)

print("TFIDF:", accuracy_score(y_test, preds))

TFIDF: 0.7677419354838709


3. **Lemmatization**

In [115]:
from nltk.stem import WordNetLemmatizer

nltk.download('wordnet')

lemmatizer = WordNetLemmatizer()

def apply_lemmatization(text):
    words = word_tokenize(str(text))
    lemmas = [lemmatizer.lemmatize(w) for w in words]
    return ' '.join(lemmas)

[nltk_data] Downloading package wordnet to /Users/ggang/nltk_data...


In [116]:
X_train_lemm = X_train.map(apply_lemmatization)
X_test_lemm  = X_test.map(apply_lemmatization)

In [117]:
vectorizer = CountVectorizer(binary=True)
X_train_lemm_vec = vectorizer.fit_transform(X_train_lemm)
X_test_lemm_vec  = vectorizer.transform(X_test_lemm)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_lemm_vec, y_train)
preds = model.predict(X_test_lemm_vec)

print("0/1:", accuracy_score(y_test, preds))

0/1: 0.7664516129032258


In [118]:
vectorizer = CountVectorizer()
X_train_lemm_vec = vectorizer.fit_transform(X_train_lemm)
X_test_lemm_vec  = vectorizer.transform(X_test_lemm)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_lemm_vec, y_train)
preds = model.predict(X_test_lemm_vec)

print("word counts:", accuracy_score(y_test, preds))

word counts: 0.7651612903225806


In [119]:
vectorizer = TfidfVectorizer()
X_train_lemm_vec = vectorizer.fit_transform(X_train_lemm)
X_test_lemm_vec  = vectorizer.transform(X_test_lemm)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_lemm_vec, y_train)
preds = model.predict(X_test_lemm_vec)

print("TFIDF:", accuracy_score(y_test, preds))

TFIDF: 0.7625806451612903


4. **Stemming + Misspellings**

In [22]:
from spellchecker import SpellChecker

spell = SpellChecker()

def fix_misspellings(text):
    words = str(text).split()
    fixed = []
    for w in words:
        correction = spell.correction(w)
        fixed.append(correction if correction else w)
    return ' '.join(fixed)

In [ ]:
X_train_spel = X_train.map(fix_misspellings)
X_train_stem_spel = X_train_spel.map(apply_stemming)
X_test_spel = X_test.map(fix_misspellings)
X_test_stem_spel  = X_test_spel.map(apply_stemming)

In [142]:
vectorizer = CountVectorizer(binary=True)
X_train_stem_spel_vec = vectorizer.fit_transform(X_train_stem_spel)
X_test_stem_spel_vec  = vectorizer.transform(X_test_stem_spel)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_stem_spel_vec, y_train)
preds = model.predict(X_test_stem_spel_vec)

print("0/1:", accuracy_score(y_test, preds))

0/1: 0.7741935483870968


In [143]:
vectorizer = CountVectorizer()
X_train_stem_spel_vec = vectorizer.fit_transform(X_train_stem_spel)
X_test_stem_spel_vec  = vectorizer.transform(X_test_stem_spel)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_stem_spel_vec, y_train)
preds = model.predict(X_test_stem_spel_vec)

print("word counts:", accuracy_score(y_test, preds))

word counts: 0.7741935483870968


In [144]:
vectorizer = TfidfVectorizer()
X_train_stem_spel_vec = vectorizer.fit_transform(X_train_stem_spel)
X_test_stem_spel_vec  = vectorizer.transform(X_test_stem_spel)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_stem_spel_vec, y_train)
preds = model.predict(X_test_stem_spel_vec)

print("TFIDF:", accuracy_score(y_test, preds))

TFIDF: 0.7793548387096774


5. **Lemmatization + Misspellings**

In [ ]:
X_train_spel = X_train.map(fix_misspellings)
X_train_lemm_spel = X_train_spel.map(apply_lemmatization)
X_test_spel = X_test.map(fix_misspellings)
X_test_lemm_spel  = X_test_spel.map(apply_lemmatization)

In [146]:
vectorizer = CountVectorizer(binary=True)
X_train_lemm_spel_vec = vectorizer.fit_transform(X_train_lemm_spel)
X_test_lemm_spel_vec  = vectorizer.transform(X_test_lemm_spel)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_lemm_spel_vec, y_train)
preds = model.predict(X_test_lemm_spel_vec)

print("0/1:", accuracy_score(y_test, preds))

0/1: 0.7651612903225806


In [147]:
vectorizer = CountVectorizer()
X_train_lemm_spel_vec = vectorizer.fit_transform(X_train_lemm_spel)
X_test_lemm_spel_vec  = vectorizer.transform(X_test_lemm_spel)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_lemm_spel_vec, y_train)
preds = model.predict(X_test_lemm_spel_vec)

print("word counts:", accuracy_score(y_test, preds))

word counts: 0.7703225806451612


In [148]:
vectorizer = TfidfVectorizer()
X_train_lemm_spel_vec = vectorizer.fit_transform(X_train_lemm_spel)
X_test_lemm_spel_vec  = vectorizer.transform(X_test_lemm_spel)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_lemm_spel_vec, y_train)
preds = model.predict(X_test_lemm_spel_vec)

print("TFIDF:", accuracy_score(y_test, preds))

TFIDF: 0.7716129032258064


6. **Stemming + Misspellings + Stop_words**

In [ ]:
X_train_spel = X_train.map(fix_misspellings)
X_train_stem_spel = X_train_spel.map(apply_stemming)
X_test_spel = X_test.map(fix_misspellings)
X_test_stem_spel  = X_test_spel.map(apply_stemming)

In [150]:
vectorizer = CountVectorizer(binary=True, stop_words='english')
X_train_stem_spel_vec = vectorizer.fit_transform(X_train_stem_spel)
X_test_stem_spel_vec  = vectorizer.transform(X_test_stem_spel)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_stem_spel_vec, y_train)
preds = model.predict(X_test_stem_spel_vec)

print("0/1:", accuracy_score(y_test, preds))

0/1: 0.7496774193548387


In [151]:
vectorizer = CountVectorizer(stop_words='english')
X_train_stem_spel_vec = vectorizer.fit_transform(X_train_stem_spel)
X_test_stem_spel_vec  = vectorizer.transform(X_test_stem_spel)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_stem_spel_vec, y_train)
preds = model.predict(X_test_stem_spel_vec)

print("word counts:", accuracy_score(y_test, preds))

word counts: 0.7496774193548387


In [152]:
vectorizer = TfidfVectorizer(stop_words='english')
X_train_stem_spel_vec = vectorizer.fit_transform(X_train_stem_spel)
X_test_stem_spel_vec  = vectorizer.transform(X_test_stem_spel)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_stem_spel_vec, y_train)
preds = model.predict(X_test_stem_spel_vec)

print("TFIDF:", accuracy_score(y_test, preds))

TFIDF: 0.7483870967741936


7. **Lemmatization + Misspellings + Stop_words**

In [ ]:
X_train_spel = X_train.map(fix_misspellings)
X_train_lemm_spel = X_train_spel.map(apply_lemmatization)
X_test_spel = X_test.map(fix_misspellings)
X_test_lemm_spel  = X_test_spel.map(apply_lemmatization)

In [154]:
vectorizer = CountVectorizer(binary=True, stop_words='english')
X_train_lemm_spel_vec = vectorizer.fit_transform(X_train_lemm_spel)
X_test_lemm_spel_vec  = vectorizer.transform(X_test_lemm_spel)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_lemm_spel_vec, y_train)
preds = model.predict(X_test_lemm_spel_vec)

print("0/1:", accuracy_score(y_test, preds))

0/1: 0.7445161290322581


In [155]:
vectorizer = CountVectorizer(stop_words='english')
X_train_lemm_spel_vec = vectorizer.fit_transform(X_train_lemm_spel)
X_test_lemm_spel_vec  = vectorizer.transform(X_test_lemm_spel)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_lemm_spel_vec, y_train)
preds = model.predict(X_test_lemm_spel_vec)

print("word counts:", accuracy_score(y_test, preds))

word counts: 0.7393548387096774


In [156]:
vectorizer = TfidfVectorizer(stop_words='english')
X_train_lemm_spel_vec = vectorizer.fit_transform(X_train_lemm_spel)
X_test_lemm_spel_vec  = vectorizer.transform(X_test_lemm_spel)

model = LogisticRegression(max_iter=1000)
model.fit(X_train_lemm_spel_vec, y_train)
preds = model.predict(X_test_lemm_spel_vec)

print("TFIDF:", accuracy_score(y_test, preds))

TFIDF: 0.7303225806451613


**Final results:**

|          | 0/1 (if word exists) | word counts | TFIDF |
|---|---|---|---|
| just tokenization | 0.7651|0.7574 | 0.7613|
| stemming | 0.7626| 0.7665|0.7677 |
| lemmatization |0.7665 |0.7652 | 0.7626|
| stemming + misspellings |0.7742 | 0.7742| 0.7794|
| lemmatization + misspellings |0.7652 | 0.7703| 0.7716|
| stemming + misspellings + stop-words | 0.7497| 0.7497| 0.7484|
| lemmatization + misspellings + stop-words |0.7445 |0.7394 | 0.7303|

8. **Custom pipeline - misspelling + stemming + vec(TFIDF with params) + logreg**

In [ ]:
X_train_spel = X_train.map(fix_misspellings)
X_test_spel  = X_test.map(fix_misspellings)
X_train_spel_stem = X_train_spel.map(apply_stemming)
X_test_spel_stem  = X_test_spel.map(apply_stemming)

In [158]:
vectorizer = TfidfVectorizer(
    ngram_range=(1, 2),
    min_df=2,
    max_df=0.9,
    sublinear_tf=True
)
X_train_vec = vectorizer.fit_transform(X_train_spel_stem)
X_test_vec  = vectorizer.transform(X_test_spel_stem)

In [162]:
model = LogisticRegression(max_iter=1000)
model.fit(X_train_vec, y_train)
preds = model.predict(X_test_vec)

print("misspelling + stemming + TFIDF(tuned) + LogReg:", accuracy_score(y_test, preds))

misspelling + stemming + TFIDF(tuned) + LogReg: 0.7806451612903226


**Similarity**

In [1]:
from sklearn.metrics.pairwise import cosine_similarity

In [14]:
texts = df.index.to_series().reset_index(drop=True)
texts = texts.astype(str)

vectorizer = TfidfVectorizer()
tfidf = vectorizer.fit_transform(texts)
print("matrix shape:", tfidf.shape)

matrix shape: (3873, 6433)


In [15]:
sim = cosine_similarity(tfidf)
print("proximity matrix:", sim.shape)

proximity matrix: (3873, 3873)


In [17]:
sim_clean = np.triu(sim, k=1)

N = 10
flat = sim_clean.flatten()
top_idx = flat.argsort()[-N:][::-1]

print(f"Top-{N} similarity tweets:\n")
for idx in top_idx:
    i, j = np.unravel_index(idx, sim.shape)
    score = sim[i, j]
    print(f"score = {score:.3f}")
    print(f"  [{i}] {texts.iloc[i]}")
    print(f"  [{j}] {texts.iloc[j]}")
    print()

Top-10 similarity tweets:

score = 1.000
  [3219] Superstar thanks so much for the kind words! Look forward to seeing you soon buddy smile.1
  [3343] Superstar thanks so much for the kind words! Look forward to seeing you soon buddy smile.2

score = 1.000
  [3219] Superstar thanks so much for the kind words! Look forward to seeing you soon buddy smile.1
  [3805] Superstar thanks so much for the kind words! Look forward to seeing you soon buddy smile.4

score = 1.000
  [3343] Superstar thanks so much for the kind words! Look forward to seeing you soon buddy smile.2
  [3805] Superstar thanks so much for the kind words! Look forward to seeing you soon buddy smile.4

score = 1.000
  [3343] Superstar thanks so much for the kind words! Look forward to seeing you soon buddy smile.2
  [3696] Superstar thanks so much for the kind words! Look forward to seeing you soon buddy smile.3

score = 1.000
  [3360] have a great Thursday! Looking forward to reading your tweets   Want this ?
  [3408] have 

**Different algorithms**

In [25]:
X_train_spel = X_train.map(fix_misspellings)
X_test_spel  = X_test.map(fix_misspellings)
X_train_prep = X_train_spel.map(apply_stemming)
X_test_prep  = X_test_spel.map(apply_stemming)

vectorizer = TfidfVectorizer(ngram_range=(1, 2), min_df=2, max_df=0.9, sublinear_tf=True)
X_train_vec = vectorizer.fit_transform(X_train_prep)
X_test_vec  = vectorizer.transform(X_test_prep)
print("features:", X_train_vec.shape[1])

features: 4054


In [26]:
from sklearn.linear_model import SGDClassifier
from sklearn.svm import LinearSVC
from sklearn.naive_bayes import MultinomialNB, ComplementNB

models = {
    "LogisticRegression": LogisticRegression(max_iter=1000, class_weight='balanced'),
    "LinearSVC":          LinearSVC(class_weight='balanced'),
    "SGDClassifier":      SGDClassifier(max_iter=1000, class_weight='balanced'),
    "MultinomialNB":      MultinomialNB(),
    "ComplementNB":       ComplementNB(),
}

results = {}
for name, m in models.items():
    m.fit(X_train_vec, y_train)
    acc = accuracy_score(y_test, m.predict(X_test_vec))
    results[name] = acc
    print(f"{name:20s}: {acc:.4f}")

LogisticRegression  : 0.7910
LinearSVC           : 0.7794
SGDClassifier       : 0.7574
MultinomialNB       : 0.7935
ComplementNB        : 0.7935
